### DISCUSSION TITLE
Is Rector able to convert PHP templates to TWIG?

### DISCUSSION THREAD

**Discussion by escopecz** [rectorphp/rector#7097] [Q&A] [answered]
Hi, the Mautic open source project is using Symfony's PHP templates but support for that was removed in Symfony 5. We are in process to update to Symfony 5 and this is one of the biggest pains to get there. I know this is a niche problem as every other project use twig so we'd have to write the rules ourselves. But I wanted to confirm with someone with more Rector experience whether it is even possible. The template files are obviously heavily infected with HTML and I know that other tools like CS fixers don't play nicely with these templates.

Here is an example of a PHP template file that needs converting to TWIG:

https://github.com/mautic/mautic/blob/4.x/app/bundles/LeadBundle/Views/Lead/lead.html.php

**Comment by TomasVotruba** [comment, marked as answer]
Hi Honza,
I love this crazy idea :) 

> I know this is a niche problem as every other project use twig so we'd have to write the rules ourselves

I disagree, there are many projects who use PHP templates and need to migrate too. It would be great addition to https://github.com/rectorphp/rector-symfony

The other way around would be very hard, as there is no easy to use Twig-AST parser.
But parsing PHP and dumping TWIG is easy. All you need is a custom printer - https://github.com/rectorphp/rector-src/blob/be530c5e2f56190f5fe1813ae4344cdaf3f75d7f/src/PhpParser/Printer/BetterStandardPrinter.php#L40

E.g. writing printer for `Echo_`:

```php
echo $value;
```

that would dump a string:

```twig
{{ value }}
```

**Reply by escopecz** [reply]
Thanks Tomas!

We'll need to change the whole `<?php echo $value; ?>` to `{{ value }}`. I guess it does not matter as the BetterStandardPrinter use regexes to match the code it's improving it seems.

I wanted to find examples of what it does to get the big picture but I'll need some more info. I couldn't google anything about it and [the tests](https://github.com/rectorphp/rector-src/blob/be530c5e2f56190f5fe1813ae4344cdaf3f75d7f/tests/PhpParser/Printer/BetterStandardPrinterTest.php) aren't helping me to understand more about it either. Is there some resource I'm missing?

**Reply by TomasVotruba** [reply]
I'm not aware of any resources that fits your needs. This is not an easy question to answer, as first you'll need to under basics of php-parser core:

* how node parsing works
* how to create a custom node, e.g. `class TwigLiteral extends Expr`, `class TwigForeach extends Foreach_` etc.
* create custom Rector rules or node traversers
* how to replace printer in php-parser/Rector for a one that allows to print custom nodes with propper Twig-like formatting

I'd start with these sources:

* https://github.com/nikic/PHP-Parser/blob/master/doc/component/Walking_the_AST.markdown
* https://github.com/nikic/PHP-Parser/blob/master/doc/component/Pretty_printing.markdown
* https://tomasvotruba.com/blog/2017/11/06/how-to-change-php-code-with-abstract-syntax-tree/#4-Steps-to-Changing-a-name

I'd also offer [a private training](https://tomasvotruba.com/trainings/) to speed this up :+1:

**Comment by effulgentsia** [comment]
Folks here might be interested in https://github.com/makinacorpus/php-twig-converter (I'm not involved with that project, just found it via a Google search). Last commit there was 7 years ago, so I don't know how much mileage you'll get from it, but it could be a helpful start nonetheless.
